pkg/genlib/generator_with_template.go

```diff
@@ -0,0 +1,529 @@
+				templateBuffer.WriteString(fieldTemplate)
+			}
+		} else {
+			fieldTemplate := fmt.Sprintf(`"%s": %s{{.%s}}%s%s`, field.Name, fieldWrap, field.Name, fieldWrap, fieldTrailer)
+			templateBuffer.WriteString(fieldTemplate)
+		}
+	}
+
+	return templateBuffer.Bytes()
+}
+
+func NewGeneratorWithTemplate(template []byte, cfg Config, fields Fields) (*GeneratorWithTemplate, error) {
+	if len(template) == 0 {
+		template = generateTemplateFromField(cfg, fields)
+	}
+
+	tokenizer := regexp.MustCompile(`([^{]*)({{\.[^}]+}})*`)
+	allIndexes := tokenizer.FindAllSubmatchIndex(template, -1)
+
+	var fieldPrefixBuffer []byte
+	var fieldPrefixPreviousN int
+
+	orderedField := make([]string, 0, len(allIndexes))
+	templateFieldMap = make(map[string][]byte, len(allIndexes))
+
+	for _, loc := range allIndexes {
+		var fieldName []byte
+		var fieldPrefix []byte
+
+		if loc[4] > -1 && loc[5] > -1 {
+			fieldName = template[loc[4]+3 : loc[5]-2]
+		}
+
+		if loc[2] > -1 && loc[3] > -1 {
+			fieldPrefix = template[loc[2]:loc[3]]
+		}
+
+		if len(fieldName) == 0 {
+			if template[fieldPrefixPreviousN] == byte(123) {
+				fieldPrefixBuffer = append(fieldPrefixBuffer, byte(123))
+			} else {
+				fieldPrefixBuffer = append(fieldPrefixBuffer, fieldPrefix...)
+			}
+		} else {
+			fieldPrefixBuffer = append(fieldPrefixBuffer, fieldPrefix...)
+			templateFieldMap[string(fieldName)] = fieldPrefixBuffer
+			orderedField = append(orderedField, string(fieldName))
+			fieldPrefixBuffer = nil
+		}
+
+		fieldPrefixPreviousN = loc[2]
+	}
+
+	trailingTemplate = fieldPrefixBuffer
+	// Preprocess the fields, generating appropriate emit functions
+	fieldMap := make(map[string]emitF)
+	for _, field := range fields {
+		if err := bindField(cfg, field, fieldMap, true); err != nil {
+			return nil, err
+		}
+	}
+
+	// Roll into slice of emit functions
+	emitFuncs := make([]emitF, 0, len(fieldMap))
+	for _, fieldName := range orderedField {
+		f := fieldMap[fieldName]
+		emitFuncs = append(emitFuncs, f)
+	}
+
+	return &GeneratorWithTemplate{emitFuncs: emitFuncs}, nil
+}
+
+func fieldValueWrapByType(field Field) string {
+	fieldType := field.Type
+	switch fieldType {
+	case FieldTypeDate, FieldTypeIP:
+		return "\""
+	case FieldTypeDouble, FieldTypeFloat, FieldTypeHalfFloat, FieldTypeScaledFloat:
+		return ""
+	case FieldTypeInteger, FieldTypeLong, FieldTypeUnsignedLong: // TODO: generate > 63 bit values for unsigned_long
+		return ""
+	case FieldTypeConstantKeyword:
+		return "\""
+	case FieldTypeKeyword:
+		return "\""
+	case FieldTypeBool:
+		return ""
+	case FieldTypeObject, FieldTypeNested, FieldTypeFlattened:
+		if len(field.ObjectType) > 0 {
+			field.Type = field.ObjectType
+		} else {
+			field.Type = FieldTypeKeyword
+		}
+		return fieldValueWrapByType(field)
+	case FieldTypeGeoPoint:
+		return "\""
+	default:
+		return "\""
+	}
+}
+
+func bindConstantKeywordWithTemplate(field Field, fieldMap map[string]emitF) error {
+	prefix := templateFieldMap[field.Name]
```

Comment:
everywhere, `bind*WithTemplate` difference from `bind*` is
```
prefix := templateFieldMap[field.Name]
```
vs
```
prefix := fmt.Sprintf("\"%s\":", field.Name)
```

we also don't write wrapping quotes, this must be taken care by the template